liquidprompt

```diff
@@ -150,6 +150,129 @@ case "$(uname)" in
     *)         LP_OS=Linux   ;;
 esac
 
+_lp_os() {
+    # Fine-grained OS detection: family, kernel, distrib, version.
+    local uname_sm kernel
+    if ! uname_sm="$(uname -s -m)"; then
+        return 1
+    fi
+    read -r kernel lp_os_arch <<<"$uname_sm"
+
+    lp_os_family=
+    lp_os_kernel=
+    lp_os_distrib=
+    lp_os_version=
+    case "$kernel" in
+        FreeBSD)
+            lp_os_family=BSD
+            lp_os_kernel=FreeBSD
+            ;;
+        DragonFly)
+            lp_os_family=BSD
+            lp_os_kernel=FreeBSD
+            ;;
+        OpenBSD)
+            lp_os_family=BSD
+            lp_os_kernel=OpenBSD
+            ;;
+        Darwin)
+            lp_os_family=BSD
+            lp_os_kernel=Darwin
+            ;;
+        SunOS)
+            lp_os_family=UNIX
+            lp_os_kernel=SunOS
+            ;;
+        CYGWIN*)
+            lp_os_family=Windows
+            lp_os_kernel=Cygwin
+            ;;
+        MSYS*)
+            lp_os_family=Windows
+            lp_os_kernel=MSYS
+            ;;
+        MINGW*)
+            lp_os_family=Windows
+            lp_os_kernel=MinGW
+            ;;
+        *) # Assume Linux
+            lp_os_kernel=Linux
+            if [[ "$kernel" == "Linux" ]]; then
+                osfamily="$(uname -o)"
+                if [[ "$osfamily" == "Android" ]]; then
+                    lp_os_family=Android
+                    return
+                fi # Android
+            fi # Linux
+            lp_os_family=GNU
+            # TODO parse HOME_URL and add an hyperlink.
+            if _lp_grep_fields "/etc/os-release" "=" "ID" "VERSION_CODENAME"; then
+                lp_os_distrib=${lp_grep_fields[_LP_FIRST_INDEX+0]-}
+                lp_os_version=${lp_grep_fields[_LP_FIRST_INDEX+1]-}
+            else
+                return "$?"
+            fi
+            ;;
+    esac
+}
+
+_lp_os_color() {
+    # Substitute lp_os_* names with icons/colored strings (from LP_MARK_OS),
+    # or hash colored versions.
+    # Join everything with LP_MARK_OS_SEP.
+    (( LP_ENABLE_OS )) || return 2
+    lp_os_color=
+    if _lp_os; then
+        local active=()
+        if (( LP_ENABLE_OS_ARCH )); then
```

Comment:
These config flag checks really should be done in `_lp_os()`, where if the flag is false, the data is never even looked up in the first place.

There here (or any other theme function), you just check if the string is not empty.

Reply:
I was thinking that this may be used as a fine-grained tool for OS-dependent features. But at the end I decided not to touch `LP_OS`, which has a slightly different scheme. Should I try to get a single set of function for those two purposes (feature management VS data display), or should I keep them separated?

Reply:
Also, I would go for having the checks in both places, given that useless calls to `_lp_substitute`/`_lp_hash_color` may be a bit slow.

Reply:
I have a different suggestion, see other comment.